src/ray/core_worker/object_interface.cc

```diff
@@ -139,92 +125,77 @@ Status CoreWorkerObjectInterface::Wait(const std::vector<ObjectID> &ids, int num
   (*results).resize(ids.size(), false);
 
   if (num_objects <= 0 || num_objects > static_cast<int>(ids.size())) {
-    return Status::Invalid("num_objects value is not valid");
+    return Status::Invalid(
+        "Number of objects to wait for must be between 1 and the number of ids.");
   }
 
   EnumUnorderedMap<StoreProviderType, std::unordered_set<ObjectID>>
       object_ids_per_store_provider;
   GroupObjectIdsByStoreProvider(ids, &object_ids_per_store_provider);
 
+  size_t total_count = 0;
+  for (const auto &entry : object_ids_per_store_provider) {
+    total_count += entry.second.size();
+  }
+  if (total_count != ids.size()) {
+    return Status::Invalid("Duplicated object IDs not supported in wait.");
+  }
+
+  // TODO(edoakes): this logic is not ideal, and will have to be addressed
+  // before we enable direct actor calls in the Python code. If we are waiting
+  // on a list of objects mixed between multiple store providers, we could
+  // easily end up in the situation where we're blocked waiting on one store
+  // provider while another actually has enough objects ready to fulfill
+  // 'num_objects'. This is partially addressed by trying them all once with
+  // a timeout of 0, but that does not address the situation where objects
+  // become available on the second store provider while waiting on the first.
```

Comment:
this might not be a big deal, as waiting from a list of mixed object ids from different store providers should be uncommon.

A possible approach is to use a while loop, and for each iteration, loop through all the store providers with a timeout, but that would add more complexity on the other hand.

Reply:
Agreed, but good to keep this in mind in case we do run into the weird behavior/can come up with a cleaner solution (I think your suggestion is ok but agree it would add complexity).